engine/unit.go

```diff
@@ -27,29 +29,46 @@ func NewUnit() *Unit {
 	return unit
 }
 
+func (u *Unit) admit() bool {
+	u.admitLock.Lock()
+	defer u.admitLock.Unlock()
+
+	select {
+	case <-u.ctx.Done():
+		return false
+	default:
+	}
+
+	u.wg.Add(1)
+	return true
+}
+
+func (u *Unit) stopAdmitting() {
+	u.admitLock.Lock()
```

Comment:
Does this change fix any existing test?
If not, could you add one that would fail without the synchronization here?

Reply:
this does not fix any existing test and the race detector won't be able to detect this race since all variables involved are thread safe.

Adding a test for this makes the code look pretty unnatural.  We would need to artificially / deterministically inject a context switch inside `admit()` and check in test using sleep / timeouts.  I'm not sure if it's worth it